### PR TITLE
Add `broadcast_sigterm_every_n_steps` to reduce SIGTERM broadcast overhead

### LINKED ISSUE #21487: forced CPU-GPU optimization at the end of every training step resulting in underutilization of GPU
I just realized that the feature implemented in this PR: #20825  causes CPU-GPU synchronization at the end of each training step, which negatively affects the GPU utilization. In particular the `torch.distributed.broadcast(sigterm_tensor, src=0)` in method: `_broadcast_sigterm_tensor` needs CPU to wait for GPU. This is not the case before adding this feature.
You can see this in Pytorch profiler below:

image

_Originally posted by @mojtababahrami in https://github.com/Lightning-AI/pytorch-lightning/issues/20825#issuecomment-3649816413_

### PR DESCRIPTION
**Pull request by c-pozzi**
## Summary

Adds a `broadcast_sigterm_every_n_steps` parameter to the Trainer that controls how often SIGTERM status is broadcast across ranks in distributed training. Default is `1` (every step), preserving current behavior.

The NCCL broadcast in `_broadcast_sigterm_tensor` (#20825) adds a fixed-cost CPU-GPU sync every step. For fast training loops this overhead is significant relative to step time. Benchmark details and measurements in #21487.

## Tradeoff

Worst-case SIGTERM detection delay is `(N-1) × step_time`. This is safe because SIGTERM grace periods are 30–120s, and users who benefit most (fast loops) have the smallest absolute delay (e.g., N=10, step=0.5ms → 4.5ms).

## Design

The parameter lives on `Trainer` following the existing `every_n` pattern (`log_every_n_steps`, `check_val_every_n_epoch`, `reload_dataloaders_every_n_epochs`) rather than `Strategy`, which holds communication mechanics, not loop frequency policy.

```python
trainer = Trainer(broadcast_sigterm_every_n_steps=10)
```

## Test plan

- [x] Validation rejects values < 1
- [x] Default value is 1
- [x] Broadcast interval logic correct for N=1, 5, 10

Closes #21487

----
📚 Documentation preview 📚: https://pytorch-lightning--21640.org.readthedocs.build/en/21640/

### CODE REVIEW

**Comment by c-pozzi** [comment]
@justusschock @mojtababahrami

**Comment by mojtababahrami** [comment]
This is a great way to handle it! Super thanks!

**Comment by c-pozzi** [comment]
I have been thinking about this and I wonder if exposing this as a user parameter is the right abstraction. The optimal value is step-time dependent which the user can't easily reason about. Would the maintainers prefer an adaptive internal approach that auto-tunes based on measured step time? Happy to rework if that approach seems valuable. @mojtababahrami @justusschock @deependujha

**Comment by mojtababahrami** [comment]
Ohh sorry I'm not a maintainer. Just a user :)

**Comment by aagaskar** [comment]
Isn't this going to cause a problem when `on_advance_end` hits
```
if not self._is_training_done and self.trainer.received_sigterm:
            raise SIGTERMException
```

The first rank will raise the exception at the end of any batch, but the others will continue and then hang.

I think a better approach would be a single check at the end of `on_advance_end`, where the existing one is. And across all ranks, the check should only run if `self._should_accumulate()` is True. That should minimize overhead since whenever gradients are accumulated there is already synchronization.

**Comment by deependujha** [comment]
Hi @aagaskar, I think you meant `when self._should_accumulate() is False`, since that’s when gradient synchronization happens across ranks.

That would make it a natural synchronization point for the SIGTERM check, ensuring all ranks observe it together and preventing the desync/hang scenario.
It also avoids introducing another argument to the `trainer` if this can be handled reliably at existing sync points.

---

Also the current sigterm check seems quite off, as it is at the starting of the training batch, and the corresponding test is quite weird as it intentially adds a sleep time, so other batches have progressed: https://github.com/Lightning-AI/pytorch-lightning/pull/20825/changes#diff-411516b6debb3ec43a284753ff0e5341932f355b4dcc8fc8684ba0c0acf318cd

**Comment by aagaskar** [comment]
> Hi @aagaskar, I think you meant `when self._should_accumulate() is False`, since that’s when gradient synchronization happens across ranks.

Yes, oops, that is what I meant!

**Comment by c-pozzi** [comment]
Thanks for the feedback @aagaskar and @deependujha.

I ran some DDP benchmarks with real forward + backward + gradient allreduce to measure the practical overhead. I tested 4 setups on 4× A10G. For this exercise I set `every_n=10`.

- `baseline`: no broadcast
- `every_step`: current Lightning behavior
- `every_n`: broadcast every 10 steps
- `on_sync_step`: broadcast only when `_should_accumulate()` is `False`

With `accumulate_grad_batches=1`, `on_sync_step` is basically the same as `every_step`, so it does not remove the overhead in the default case. For medium workloads the overhead is ~1.5–2%, but for very fast sub-ms loops it gets much larger.

With `accumulate_grad_batches > 1`, `on_sync_step` works really well and brings the overhead close to zero, since the check only happens on the natural gradient sync step.

So my takeaway is:
- the sync-point approach seems like the better design from a correctness/API perspective,
- but it does not seem to  fully address the fast-loop overhead when `accumulate_grad_batches=1`,
- while it does help a lot when gradients are accumulated.

**Results — `accumulate_grad_batches=1` (default)**

| Workload | Step time | every_step | on_sync_step | every_n |
|---|---:|---:|---:|---:|
| small_linear (512) | 0.87 ms | +14.4% | +14.2% | -1.1% |
| medium_linear (2048) | 8.83 ms | +1.5% | +1.7% | +0.1% |
| small_mlp (1024) | 6.98 ms | +1.9% | +1.8% | -0.0% |

**Results — `accumulate_grad_batches=4`**

| Workload | Step time | every_step | on_sync_step | every_n |
|---|---:|---:|---:|---:|
| small_linear (512) | 0.40 ms | +37.4% | -4.2% | -2.9% |
| medium_linear (2048) | 0.42 ms | +32.3% | -0.6% | +1.4% |
| small_mlp (1024) | 0.77 ms | +17.0% | -0.2% | +0.8% |

Willing to rework this PR around `on_advance_end`, triggering the SIGTERM broadcast/check only when `_should_accumulate()` is `False`, and drop the new Trainer argument if that is the preferred direction.

I can also share the benchmark script if useful.

**Comment by deependujha** [comment]
Hi @c-pozzi, thanks for the great work. I’m currently exploring why even we need this communication at all, and trying to fix how it was originally supposed to work (including reverting previously merged pr)
